File: website/lib/litellm.ts
// LiteLLM admin API: mint per-user virtual keys and read their spend.
// Uses the master key server-side only.

const BASE = process.env.LITELLM_URL!;
const MASTER = process.env.LITELLM_MASTER_KEY!;

// Free-tier model allowlist now includes parakeet-default (which chains
// through Groq → Cerebras → Cloudflare → OpenRouter free tiers) so
// hosted users get working inference at zero cost to us.
const TIER_MODELS: Record<string, string[]> = {
  free: ["parakeet-groq", "parakeet-default"],
  pro: ["parakeet-groq", "parakeet-claude", "parakeet-gpt", "parakeet-gemini", "parakeet-default"],
};

const TIER_BUDGET: Record<string, number> = {
  free: 1, // USD/month — generous since free-tier APIs cost us $0
  pro: 15,
};

function headers() {
  return {
    Authorization: `Bearer ${MASTER}`,
    "Content-Type": "application/json",
  };
}

/** Create a virtual key for a user with a monthly budget for their tier. */
export async function mintKey(userId: string, tier: string): Promise<string> {
  const res = await fetch(`${BASE}/key/generate`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify({
      models: TIER_MODELS[tier] ?? TIER_MODELS.free,
      max_budget: TIER_BUDGET[tier] ?? TIER_BUDGET.free,
      budget_duration: "30d",
      metadata: { user_id: userId },
    }),
  });
  if (!res.ok) throw new Error(`LiteLLM key/generate failed: ${res.status}`);
  const data = await res.json();
  return data.key as string;
}

/** Spend + budget for a key, for the dashboard credit meter. */
export async function getSpend(
  key: string
): Promise<{ spend: number; maxBudget: number }> {
  const res = await fetch(`${BASE}/key/info?key=${encodeURIComponent(key)}`, {
    headers: headers(),
  });
  if (!res.ok) return { spend: 0, maxBudget: 0 };
  const data = await res.json();
  const info = data.info ?? data;
  return {
    spend: Number(info.spend ?? 0),
    maxBudget: Number(info.max_budget ?? 0),
  };
}

File: webapp/webapp/services/litellm.py
"""LiteLLM admin API: mint per-user virtual keys and read spend. Server-only
— uses the master key. Mirrors website/lib/litellm.ts."""
from __future__ import annotations

import os

import httpx

TIER_BUDGET = {"free": 1.0, "pro": 15.0}

# Hosted model aliases exposed to each tier. These are the Wave 2 aliases
# (backend/litellm-config.yaml doesn't define them yet — that's tracked
# separately and does not block minting/updating keys against them here).
TIER_MODELS = {
    "free": ["parakeet-groq"],
    "pro": ["parakeet-groq", "parakeet-claude", "parakeet-gpt", "parakeet-gemini"],
}


def _base_url() -> str:
    url = os.environ.get("LITELLM_URL")
    if not url:
        raise RuntimeError("LITELLM_URL is not set")
    return url.rstrip("/")


def _headers() -> dict[str, str]:
    master_key = os.environ.get("LITELLM_MASTER_KEY")
    if not master_key:
        raise RuntimeError("LITELLM_MASTER_KEY is not set")
    return {"Authorization": f"Bearer {master_key}", "Content-Type": "application/json"}


def mint_key(user_id: str, tier: str) -> str:
    """Create a virtual key for a user with a monthly budget for their tier."""
    response = httpx.post(
        f"{_base_url()}/key/generate",
        headers=_headers(),
        json={
            "models": TIER_MODELS.get(tier, TIER_MODELS["free"]),
            "max_budget": TIER_BUDGET.get(tier, TIER_BUDGET["free"]),
            "budget_duration": "30d",
            "metadata": {"user_id": user_id},
        },
        timeout=15.0,
    )
    if response.status_code >= 400:
        raise RuntimeError(f"LiteLLM key/generate failed: {response.status_code}")
    return response.json()["key"]


def update_key_budget(key: str, tier: str) -> None:
    """Change max_budget (and the model allowlist) for an *existing* key
    without reissuing it. Used on subscription upgrade/downgrade so the key
    already saved in the user's desktop config.env keeps working."""
    response = httpx.post(
        f"{_base_url()}/key/update",
        headers=_headers(),
        json={
            "key": key,
            "models": TIER_MODELS.get(tier, TIER_MODELS["free"]),
            "max_budget": TIER_BUDGET.get(tier, TIER_BUDGET["free"]),
        },
        timeout=15.0,
    )
    if response.status_code >= 400:
        raise RuntimeError(f"LiteLLM key/update failed: {response.status_code}")


def get_spend(key: str) -> tuple[float, float]:
    """Return (spend, max_budget) for a key, for the dashboard credit meter."""
    response = httpx.get(
        f"{_base_url()}/key/info",
        params={"key": key},
        headers=_headers(),
        timeout=15.0,
    )
    if response.status_code >= 400:
        return 0.0, 0.0
    data = response.json()
    info = data.get("info", data)
    return float(info.get("spend", 0) or 0), float(info.get("max_budget", 0) or 0)
